File: ios/SalmonTrollingAssistant/SalmonTrollingAssistant/Presentation/ViewModels/CatchLoggingViewModel.swift
import Foundation
import Combine

class CatchLoggingViewModel: ObservableObject {
    private let catchLoggingService: CatchLoggingService
    private var cancellables = Set<AnyCancellable>()
    
    @Published var catchHistory: [CatchData] = []
    @Published var selectedCatch: CatchData?
    @Published var catchLoggingState: CatchLoggingState = .initial
    
    init(catchLoggingService: CatchLoggingService) {
        self.catchLoggingService = catchLoggingService
        
        // Subscribe to catch history changes
        catchLoggingService.catchHistory
            .map { catches in
                catches.sorted { $0.timestamp > $1.timestamp }
            }
            .assign(to: \.catchHistory, on: self)
            .store(in: &cancellables)
    }
    
    func logCatch(
        locationId: String,
        species: FishSpecies,
        size: Double?,
        weight: Double?,
        equipmentUsed: [String],
        weatherConditionsId: String?,
        tideConditionsId: String?,
        notes: String?,
        photoUrls: [String] = []
    ) {
        catchLoggingState = .loading
        
        let catchData = CatchData(
            id: UUID().uuidString,
            timestamp: Date(),
            locationId: locationId,
            species: species,
            size: size,
            weight: weight,
            equipmentUsed: equipmentUsed,
            weatherConditionsId: weatherConditionsId,
            tideConditionsId: tideConditionsId,
            notes: notes,
            photoUrls: photoUrls
        )
        
        Task {
            let result = await catchLoggingService.logCatch(catchData)
            
            await MainActor.run {
                switch result {
                case .success:
                    catchLoggingState = .success
                case .failure(let error):
                    catchLoggingState = .error(error.localizedDescription)
                }
            }
        }
    }
    
    func updateCatch(_ catchData: CatchData) {
        catchLoggingState = .loading
        
        Task {
            let result = await catchLoggingService.updateCatch(catchData)
            
            await MainActor.run {
                switch result {
                case .success:
                    catchLoggingState = .success
                case .failure(let error):
                    catchLoggingState = .error(error.localizedDescription)
                }
            }
        }
    }
    
    func deleteCatch(catchId: String) {
        catchLoggingState = .loading
        
        Task {
            let result = await catchLoggingService.deleteCatch(catchId: catchId)
            
            await MainActor.run {
                switch result {
                case .success:
                    catchLoggingState = .success
                case .failure(let error):
                    catchLoggingState = .error(error.localizedDescription)
                }
            }
        }
    }
    
    func addPhotoCatch(catchId: String, photoUrl: String) {
        catchLoggingState = .loading
        
        Task {
            let result = await catchLoggingService.addPhotoCatch(catchId: catchId, photoUrl: photoUrl)
            
            await MainActor.run {
                switch result {
                case .success:
                    catchLoggingState = .success
                case .failure(let error):
                    catchLoggingState = .error(error.localizedDescription)
                }
            }
        }
    }
    
    func removePhotoCatch(catchId: String, photoUrl: String) {
        catchLoggingState = .loading
        
        Task {
            let result = await catchLoggingService.removePhotoCatch(catchId: catchId, photoUrl: photoUrl)
            
            await MainActor.run {
                switch result {
                case .success:
                    catchLoggingState = .success
                case .failure(let error):
                    catchLoggingState = .error(error.localizedDescription)
                }
            }
        }
    }
    
    func selectCatch(catchId: String) {
        Task {
            let result = await catchLoggingService.getCatchById(catchId: catchId)
            
            await MainActor.run {
                switch result {
                case .success(let catchData):
                    selectedCatch = catchData
                case .failure:
                    selectedCatch = nil
                }
            }
        }
    }
    
    func clearSelectedCatch() {
        selectedCatch = nil
    }
    
    func resetCatchLoggingState() {
        catchLoggingState = .initial
    }
    
    enum CatchLoggingState: Equatable {
        case initial
        case loading
        case success
        case error(String)
        
        static func == (lhs: CatchLoggingState, rhs: CatchLoggingState) -> Bool {
            switch (lhs, rhs) {
            case (.initial, .initial):
                return true
            case (.loading, .loading):
                return true
            case (.success, .success):
                return true
            case (.error(let lhsMessage), .error(let rhsMessage)):
                return lhsMessage == rhsMessage
            default:
                return false
            }
        }
    }
}

File: android/SalmonTrollingAssistant/app/src/main/java/com/example/salmontrollingassistant/presentation/viewmodels/CatchLoggingViewModel.kt
package com.example.salmontrollingassistant.presentation.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.salmontrollingassistant.domain.model.CatchData
import com.example.salmontrollingassistant.domain.model.FishSpecies
import com.example.salmontrollingassistant.domain.service.CatchLoggingService
import com.example.salmontrollingassistant.domain.service.LocationService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.util.UUID
import javax.inject.Inject

@HiltViewModel
class CatchLoggingViewModel @Inject constructor(
    private val catchLoggingService: CatchLoggingService,
    private val locationService: LocationService
) : ViewModel() {
    
    private val _catchHistory = MutableStateFlow<List<CatchData>>(emptyList())
    val catchHistory: StateFlow<List<CatchData>> = _catchHistory.asStateFlow()
    
    private val _selectedCatch = MutableStateFlow<CatchData?>(null)
    val selectedCatch: StateFlow<CatchData?> = _selectedCatch.asStateFlow()
    
    private val _catchLoggingState = MutableStateFlow<CatchLoggingState>(CatchLoggingState.Initial)
    val catchLoggingState: StateFlow<CatchLoggingState> = _catchLoggingState.asStateFlow()
    
    init {
        loadCatchHistory()
    }
    
    private fun loadCatchHistory() {
        viewModelScope.launch {
            catchLoggingService.getCatchHistory().collectLatest { catches ->
                _catchHistory.value = catches.sortedByDescending { it.timestamp }
            }
        }
    }
    
    fun logCatch(
        locationId: String,
        species: FishSpecies,
        size: Double?,
        weight: Double?,
        equipmentUsed: List<String>,
        weatherConditionsId: String?,
        tideConditionsId: String?,
        notes: String?,
        photoUrls: List<String> = emptyList()
    ) {
        viewModelScope.launch {
            _catchLoggingState.value = CatchLoggingState.Loading
            
            val catchData = CatchData(
                id = UUID.randomUUID().toString(),
                timestamp = System.currentTimeMillis(),
                locationId = locationId,
                species = species,
                size = size,
                weight = weight,
                equipmentUsed = equipmentUsed,
                weatherConditionsId = weatherConditionsId,
                tideConditionsId = tideConditionsId,
                notes = notes,
                photoUrls = photoUrls
            )
            
            val result = catchLoggingService.logCatch(catchData)
            
            _catchLoggingState.value = if (result.isSuccess) {
                CatchLoggingState.Success
            } else {
                CatchLoggingState.Error(result.exceptionOrNull()?.message ?: "Unknown error")
            }
        }
    }
    
    fun updateCatch(catchData: CatchData) {
        viewModelScope.launch {
            _catchLoggingState.value = CatchLoggingState.Loading
            
            val result = catchLoggingService.updateCatch(catchData)
            
            _catchLoggingState.value = if (result.isSuccess) {
                CatchLoggingState.Success
            } else {
                CatchLoggingState.Error(result.exceptionOrNull()?.message ?: "Unknown error")
            }
        }
    }
    
    fun deleteCatch(catchId: String) {
        viewModelScope.launch {
            _catchLoggingState.value = CatchLoggingState.Loading
            
            val result = catchLoggingService.deleteCatch(catchId)
            
            _catchLoggingState.value = if (result.isSuccess) {
                CatchLoggingState.Success
            } else {
                CatchLoggingState.Error(result.exceptionOrNull()?.message ?: "Unknown error")
            }
        }
    }
    
    fun addPhotoCatch(catchId: String, photoUri: String) {
        viewModelScope.launch {
            _catchLoggingState.value = CatchLoggingState.Loading
            
            val result = catchLoggingService.addPhotoCatch(catchId, photoUri)
            
            _catchLoggingState.value = if (result.isSuccess) {
                CatchLoggingState.Success
            } else {
                CatchLoggingState.Error(result.exceptionOrNull()?.message ?: "Unknown error")
            }
        }
    }
    
    fun removePhotoCatch(catchId: String, photoUri: String) {
        viewModelScope.launch {
            _catchLoggingState.value = CatchLoggingState.Loading
            
            val result = catchLoggingService.removePhotoCatch(catchId, photoUri)
            
            _catchLoggingState.value = if (result.isSuccess) {
                CatchLoggingState.Success
            } else {
                CatchLoggingState.Error(result.exceptionOrNull()?.message ?: "Unknown error")
            }
        }
    }
    
    fun selectCatch(catchId: String) {
        viewModelScope.launch {
            val result = catchLoggingService.getCatchById(catchId)
            if (result.isSuccess) {
                _selectedCatch.value = result.getOrNull()
            }
        }
    }
    
    fun clearSelectedCatch() {
        _selectedCatch.value = null
    }
    
    fun resetCatchLoggingState() {
        _catchLoggingState.value = CatchLoggingState.Initial
    }
    
    sealed class CatchLoggingState {
        object Initial : CatchLoggingState()
        object Loading : CatchLoggingState()
        object Success : CatchLoggingState()
        data class Error(val message: String) : CatchLoggingState()
    }
}